modules/map/tools/map_datachecker/server/alignment.h

```diff
@@ -0,0 +1,141 @@
+
+namespace apollo {
+namespace hdmap {
+
+typedef struct BadOrGoodPoseInfo {
+  BadOrGoodPoseInfo() : start_time(-1.0), end_time(-1.0), pose_count(0) {}
+  double start_time;
+  double end_time;
+  int pose_count;
+} BadOrGoodPoseInfo;
+
+class Alignment {
+ public:
+  explicit Alignment(std::shared_ptr<JSonConf> sp_conf)
+      : return_state_(ErrorCode::SUCCESS),
+        sp_conf_(sp_conf),
+        sp_good_pose_info_(std::make_shared<BadOrGoodPoseInfo>()),
+        sp_bad_pose_info_(std::make_shared<BadOrGoodPoseInfo>()) {}
+
+  virtual ~Alignment() {}
+  virtual ErrorCode Process(const std::vector<FramePose>& poses) = 0;
+  virtual void Reset() = 0;
+
+  virtual double GetProgress() const { return progress_; }
+
+  virtual void SetStartTime(double start_time) { start_time_ = start_time; }
+
+  virtual void SetEndTime(double end_time) { end_time_ = end_time; }
+
+  virtual void UpdateBadPoseInfo(const FramePose& pose) {
+    UpdatePoseInfo(pose, sp_bad_pose_info_);
+  }
+
+  virtual void ClearBadPoseInfo() { ClearPoseInfo(sp_bad_pose_info_); }
+
+  virtual void UpdateGoodPoseInfo(const FramePose& pose) {
+    UpdatePoseInfo(pose, sp_good_pose_info_);
+  }
+
+  virtual void ClearGoodPoseInfo() { ClearPoseInfo(sp_good_pose_info_); }
+
+  virtual bool IsGoodPose(const std::vector<FramePose>& poses,
+                            int pose_index) {
+    if (pose_index <= 0 || pose_index >= static_cast<int>(poses.size())) {
+      AINFO << "params error. poses size:" << poses.size()
+            << ",pose_index:" << pose_index;
+      return false;
+    }
+
+    unsigned int position_type = poses[pose_index].position_type;
+    float diff_age = poses[pose_index].diff_age;
+    double local_std = poses[pose_index].local_std;
+
+    if (sp_conf_->position_type_range.find(position_type) !=
+            sp_conf_->position_type_range.end() &&
+        diff_age >= sp_conf_->diff_age_range.first &&
+        diff_age <= sp_conf_->diff_age_range.second &&
+        local_std <= sp_conf_->local_std_upper_limit) {
+      return true;
+    }
+    return false;
+  }
+
+  ErrorCode GetReturnState() const { return return_state_; }
+
+ protected:
+  void UpdatePoseInfo(const FramePose& pose,
+                        std::shared_ptr<BadOrGoodPoseInfo> sp_pose_info) {
+    BadOrGoodPoseInfo& pose_info = *sp_pose_info;
+    if (pose_info.pose_count == 0) {
+      pose_info.start_time = pose.time_stamp;
+      ++pose_info.pose_count;
+      AINFO << "update start time: " << std::to_string(pose_info.start_time)
+            << ",pose count: " << pose_info.pose_count;
+    } else {
+      pose_info.end_time = pose.time_stamp;
+      ++pose_info.pose_count;
+      AINFO << "update start time: " << std::to_string(pose_info.start_time)
+            << ",pose count: " << pose_info.pose_count;
+    }
+  }
+
+  void ClearPoseInfo(std::shared_ptr<BadOrGoodPoseInfo> sp_pose_info) {
+    BadOrGoodPoseInfo& pose_info = *sp_pose_info;
```

Comment:
Before using a shared_pointer, it's better to check whether the pointer is nullptr.